python/ray/data/_internal/datasource/parquet_datasink.py

```diff
@@ -194,12 +203,57 @@ def _write_parquet_files(
             self.mode, "overwrite_or_ignore"
         )
 
-        # Set default row group size if not provided. Defaults are set by pyarrow.
-        min_rows_per_group = row_group_size if row_group_size else 0
-        max_rows_per_group = row_group_size if row_group_size else 1024 * 1024
-
         basename_template = self._get_basename_template(filename, write_uuid)
 
+        # Default to 0 set by pyarrow.dataset.write_dataset
+        ### From docs:
+        # max_rows_per_file : int, default 0
+        # Maximum number of rows per file. If greater than 0 then this will
+        # limit how many rows are placed in any single file. Otherwise there
+        # will be no limit and one file will be created in each output
+        # directory unless files need to be closed to respect max_open_files
+        max_rows_per_file = self.max_rows_per_file or 0
+
+        # ---------------------------------------------------------------------------
+        # Split each PyArrow table into chunks that respect {min,max}_rows_per_file
+        # and the target block size detected by _get_max_chunk_size().
+        # ---------------------------------------------------------------------------
+        # ── Derive a *global* block‑row budget from all incoming batches ───
+        #      (smallest estimate is safest, avoids blowing RAM on large columns)
+        min_rows_per_file = self.min_rows_per_file  # can be None
+        min_block_rows: Optional[int] = None
+        for tbl in tables:
+            estimated_block_size = _get_max_chunk_size(
+                tbl, self._data_context.target_max_block_size
+            )
+            if estimated_block_size:
+                min_block_rows = (
+                    estimated_block_size
+                    if min_block_rows is None
+                    else min(min_block_rows, estimated_block_size)
+                )
+        if min_block_rows is None:
+            min_block_rows = 1_024 * 1_024  # fall back to Arrow default
+
+        # ── Decide row‑group sizing so each *file* lands within the         ─
+        #   [min_rows_per_file, max_rows_per_file] band *and* respects RAM.  ─
+        if row_group_size is None:
+            min_rows_per_group = min_rows_per_file or 0
+            max_rows_per_group = min_block_rows
+            if max_rows_per_file > 0:
+                max_rows_per_group = min(max_rows_per_group, max_rows_per_file)
```

Comment:
Why do we still need all of that? Instead of just passing min/max to write_to_dataset